rep-0151.rst

```diff
@@ -0,0 +1,151 @@
+REP: 151
+Title: Python 3 Support
+Author: Dirk Thomas
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 23-Dec-2017
+
+Outline
+=======
+
+#. Abstract_
+#. Motivation_
+#. Rationale_
+#. Proposal_
+#. Requirements_
+#. References_
+#. Copyright_
+
+Abstract
+========
+
+This REP describes the path for switching from Python 2 to Python 3 in ROS 1.
+
+Motivation
+==========
+
+As described in the Python release schedule [1]_ Python 2.7 will only receive
+bugfix releases until 2020.
+As a consequence Ubuntu is trying to demote Python 2.7 into the `universe`
+repository for 18.04 since they don't want to guarantee support for 5 years.
+While it is unlikely that Python 2.7 will be removed from the archives entirely
+it should not be relied on past its EOL date.
+Therefore ROS 1 needs to begin moving to Python 3 in order to make the
+transition as smooth as possible.
+
+Rationale
+=========
+
+The goal for a future ROS 1 distribution to support Python 3 should be achieved
+no later than for ROS O-turtle (May 2020).
+Since O-turtle is a LTS release it is strongly desired to already support
+Python 3 in the ROS N-turtle release to allow enough testing time before the
+mandatory switch.
+
+In general a single Python version must be used since different versions can't
+be mixed within a single ROS distribution.
+E.g. a package A using Python 3 can't use the Python modules from package B if
+that was built with/for Python 2.
+
+In order to support Python 3 it is mandatory that the testing infrastructure
+(namely `devel` and `PR` jobs) can run the build and tests with Python 3.
+Since each of these jobs require Debian packages of the dependencies to be
+present being able to test for a specific Python version requires Debian
+packages for this Python version to be present.
+The alternative of building all dependencies from-source in these jobs doesn't
+scale and is therefore not feasible since the time necessary for a single build
+gets increasingly high the further up in the stack the package is.
+
+Updating all ROS packages which contain Python code to support Python 2
+as well as 3 is a significant effort.
+Therefore it is expected that this will take a while and will require the ROS
+community to share the necessary workload.
+
+Proposal
+========
+
+For a smooth transition the following timeline is being proposed:
+
+ * ROS O-turtle (May 2020, LTS):
+
+   * The only Debian packages provided are using Python 3.
+   * From-source builds are using Python 3 by default.
+   * All ROS packages should support Python 3.
+   * Python 2 support is not required anymore.
+
+ * ROS N-turtle (May 2019):
+
+   * The set of recommended Debian packages is using Python 3.
+   * There is an alternative set of Debian packages which is using Python 2,
+     this can be used in case some packages don't support Python 3 yet.
```

Comment:
It would be valuable to more explicitly define how the "alternative" packages will be made available in N and M. Some options may include:

- Alternative component: `deb http://packages.ros.org/ros/ubuntu bionic main-py2`
- Alternative repo: `deb http://packages.ros.org/ros/ubuntu-py2 bionic main`
- Alternative suite + version number shenanigans (since suites share a common name+version space): `deb http://packages.ros.org/ros/ubuntu bionic-py2 main`
- Package name prefix/suffix.

It's also not clear if the "alternative" packages can be co-installed with the default ones, which will only be possible if they have different names. If that's not the case, there should be a plan to offer scripts which debootstrap the alternative config into a chroot so that users have a not-insane workflow for at least building and running tests locally.

Reply:
I think it is reasonable to expect that the Python 2 debs as well as the Python 3 debs can be installed side-by-side. I added this note in 8a575ffacae943fb0f3d3643e776361421803b65.

I don't think that the implementation details how this is achieved needs to be specified in this REP though. For the goal of side-by-side installation it certainly requires the Debian package name to be different as well as the installed files to be in a different location.

Reply:
👍